File: 24.swap-nodes-in-pairs.java
/*
 * @lc app=leetcode id=24 lang=java
 *
 * [24] Swap Nodes in Pairs
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode be4Head = new ListNode(0);
        be4Head.next = head;
        ListNode prevPtr = be4Head, curPtr = prevPtr.next, nextPtr = curPtr.next;
        while (nextPtr != null) {
            prevPtr.next = nextPtr;
            curPtr.next = nextPtr.next;
            nextPtr.next = curPtr;
            prevPtr = curPtr;
            curPtr = curPtr.next;
            if (curPtr == null) break;
            nextPtr = curPtr.next;
        }
        return be4Head.next;
    }
}

File: go/24.swap-nodes-in-pairs.go
/*
 * @lc app=leetcode id=24 lang=golang
 *
 * [24] Swap Nodes in Pairs
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}
	ahead := &ListNode{0, head}
	prev, cur, next := ahead, head, head.Next
	for next != nil {
		tmpNext := next.Next
		prev.Next, cur.Next, next.Next = next, tmpNext, cur
		if tmpNext == nil {
			break
		}
		prev, cur, next = cur, tmpNext, tmpNext.Next
	}
	return ahead.Next
}
// @lc code=end

